feat!: modulestore_migrator

kdmccormick opened the pull request:
## Description

This is a work-in-progress implementation of:
* https://github.com/openedx/frontend-app-authoring/issues/1681

for the Content Libraries Relaunch project.

It mostly follows the decisions captured [in this ADR](https://github.com/openedx/edx-platform/pull/36545). It is a rework of the [previously-merged import_from_modulestore app](https://github.com/openedx/edx-platform/pull/36515).

This includes:

* The migration task itself
* A simple Python API which calls the task
* A django admin interface

This needs:
- [x] More manual testing
- [x] Several kinks worked out... see my `@@TODO` comments. Notable issues:
  - [x] ~Need to decide if we like the target block slug structure (`Org__Course__Run__SourceBlockSlug`)~ changed
  - [x] ~Need to decide if we can assume the target block slug structure, or if we should be looking into previous migrations when deciding whether to create an entity or update an existing one~ changed
  - [x] ~Need to fix collections; currently just works with importing into the library~ Moved to follow-up
  - [x] Need to fix the change_log_record association
- [x] Unit tests
- [x] REST API
- [x] More comments
- [x] ~A migration to delete the import_from_modulestore app?~ done separately

## Supporting information

 ### The admin interface

**Go to `<CMS_ROOT>/admin/modulestore_migrator/modulestoresource`**

**Add a ModulestoreSource** with the key of the course you'd like to import. You only need to do this once per course, even if you migrate from it multiple times.

<img width="1190" alt="Screenshot 2025-06-11 at 6 02 42 PM" src="https://github.com/user-attachments/assets/cac39e41-b27f-485d-8abf-0bb87825b7b1" />

**Select the source**, choose "Start migration for selected sources", specify a target library key, whether to replace existing components, and what hierarchy level to migrate up to.

<img width="1190" alt="Screenshot 2025-06-11 at 6 00 26 PM" src="https://github.com/user-attachments/assets/ff9e33d3-6195-468a-8858-90ee6ab47933" />

**Hit Go** and observe that your migration(s) have started

<img width="1190" alt="Screenshot 2025-06-11 at 6 04 31 PM" src="https://github.com/user-attachments/assets/cd840ed6-0edb-4992-8174-48ea09d8dac3" />

**Check up on the status** by clicking into your Modulestore Source. You can see past migrations, too.

<img width="1190" alt="Screenshot 2025-06-11 at 6 05 13 PM" src="https://github.com/user-attachments/assets/097fafc0-c6f3-4153-a8af-cb9b29037a7b" />

**See the result** by clicking "Change" on your Migration. You can see the blocks that were migrated in.

<img width="1190" alt="Screenshot 2025-06-11 at 6 06 27 PM" src="https://github.com/user-attachments/assets/5c37fe9d-f469-47eb-8154-d67c921ae231" />

Commit:
feat: redesign import task + api/helpers tests adaptation

Commit:
feat: add simple REST API

Commit:
refactor: add API endpoint secs to view docs, updated tests & fix various lint errors

Commit:
refactor: simplify staged content in helpers tests & fix import order in tasks.py

Commit:
fix: add default value for composition_level to not break old instances

kdmccormick committed:
feat!: modulestore_migrator (WIP)

kdmccormick commented:
@cmltaWt0 Here is the PR I am working off of. It reworks the import_from_modulstore app into modulestore_migrator. There are still some parts I need to polish off tomorrow (Monday) before I can hand it off, but feel free to start familiarizing yourself with it.

kdmccormick commented:
@cmltaWt0 OK, I believe the new modulestore_migrator code for models.py, api.py, and tasks.py is all complete. What I will need RaccoonGang's help with is porting admin.py, rest_api/v0/views.py, and all the unit tests over from import_from_modulestore into the new modulestore_migrator.

I know there are a lot of changes here, and it probably will not be clear why I am insisting on such an extensive rework. The short answer is that, over time, this new app will serve as the basis for the migration of *all* courses and libraries into learning core; in order to support that, I had to rethink the models. Tomorrow I will work on a more thorough explanation and an overview of the new code.

cmltaWt0 commented:
> @cmltaWt0 OK, I believe the new modulestore_migrator code for models.py, api.py, and tasks.py is all complete. What I will need RaccoonGang's help with is porting admin.py, rest_api/v0/views.py, and all the unit tests over from import_from_modulestore into the new modulestore_migrator.
> 
> I know there are a lot of changes here, and it probably will not be clear why I am insisting on such an extensive rework. The short answer is that, over time, this new app will serve as the basis for the migration of _all_ courses and libraries into learning core; in order to support that, I had to rethink the models. Tomorrow I will work on a more thorough explanation and an overview of the new code.

We are working on it - starting from tests.
New changes will be provided as a PRs to you branch. Will it be convenient?

kdmccormick commented:
> New changes will be provided as a PRs to you branch. Will it be convenient?

Yes, perfect

kdmccormick committed:
feat: add admin interface

kdmccormick commented:
@cmltaWt0 I've pushed an admin interface for starting migrations for given sources (i.e., courses):

<img width="830" alt="Screenshot 2025-06-11 at 1 33 11 PM" src="https://github.com/user-attachments/assets/b7c280e5-081d-42a5-b276-97fc26bf3f07" />



and set up admin page such that it displays the migrations and blocks associated with a source:

<img width="830" alt="Screenshot 2025-06-11 at 1 31 30 PM" src="https://github.com/user-attachments/assets/1f92a72a-0021-463c-ac59-7107ec470cae" />

kdmccormick commented:
@cmltaWt0 The migration task seems to be broken currently. I'll work on fixing that, and then work on some more documentation & explanation, and then I'll leave it in your hands.

kdmccormick committed:
feat: better task debug info on the admin interface

kdmccormick committed:
fix: fixes to tasks and admin -- happy path works now

kdmccormick commented:
@cmltaWt0 OK, the happy path is working on the migration task now. I've updated the description with a quick guide on how to use the admin interface. I'll provide some more documentation tomorrow and then fully hand it off to RG.

@bradenmacdonald , here is an admin action you can use to migrate course content into Learning Core. It still has some bugs on the edge cases and could use better progress tracking, but I think it'll be good enough for the purposes of making test data for the prototype. It goes up to the Section level. As I've mentioned, it could be adjusted to go up to OutlineRoot if that would be helpful or nice for the demo.

bradenmacdonald commented:
@kdmccormick Great! Q: Is this also going to be the basis of importing legacy libraries into new libraries, or is that a totally separate migration workflow?

kdmccormick commented:
Yep, this is the basis for legacy libraries, too. It might actually work already, haven't tried.

There is a field in the model, forwarded_by, that is meant to handle the reference forwarding. Currently it's always null since we don't need it for the import api.

Commit:
feat: add REST API for task creation & status retrieve

Commit:
test: cover new changes with tests

Commit:
test: add collection_slug case

Commit:
refactor: remove ContentLibrary & LearningPackage factories

Commit:
refactor: fix docstring style, remove redundant comments & clean a little bit

kdmccormick committed:
fix: Create ContainerVersion subclassses too

This commit stops this error:

   ContainerVersion.unitversion.RelatedObjectDoesNotExist: ContainerVersion has no unitversion.

from happening when accessing migrated content.

You will have to re-migrate the content to fix. You may even have to discard the
original broken migrated content.

kdmccormick committed:
fix: drop all the models to un-bork modulestoreblocksource

kdmccormick committed:
fix: recreate models with missing db constraint

we were missing the unique_together on modulestoreblocksource,
meaning we had multiple modulestoreblocksources for the
same legacy usage key.

also, replace bulk_create with a series of saves to un-bork
forwarding for now.

also, this renames forwarded_by -> forwarded

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10455855163.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10455855163-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10455855163-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10455855163-tutor-requirements.txt)

kdmccormick committed:
refactor: remove unnecessary non-Noneness check after get_library

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10467620409.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10467620409-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10467620409-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10467620409-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144741.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144741-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144741-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144741-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144745.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144745-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144745-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10468144745-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879908.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879908-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879908-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879908-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36873-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36873-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879820-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879820-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10472879820-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659215.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659215-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659215-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659215-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment successful 🚀
🎓 [LMS](https://pr-36873-7daceb.sandboxes.opencraft.hosting)
📝 [Studio](https://studio.pr-36873-7daceb.sandboxes.opencraft.hosting)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659211-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659211-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-10506659211-tutor-requirements.txt)

Commit:
feat: implement TODOs & fix tests

cmltaWt0 committed:
feat: implement _create_migration_artifacts_in_bulk

Left the incremental implementation to compare.

TODO:
- remove one of the competitive versions
- discuss the better option
- amend this commit and clear commin history

Performance:
- incremental for 39 Blocks 0.06 sec
- bulk for 39 Blocks 0.006 sec

cmltaWt0 committed:
feat: implement Handling duplicates in modulestore migrator

TODO:
- remove extensive loggins, it's currently required
  for debugging purpuse
- Update tests regarding the new slugification logic
  and according to new field
- update API
- debug and fix performance issue during authoring_api.add_to_collection
- refactor existing_source_to_target_keys by moving it and
  some common arguments from _migrate_node to MigrationContext
- test API and remove unused fields

cmltaWt0 committed:
feat: implement Containers restoring

Changes:
- refactor slugification
- restore a deleted container

cmltaWt0 committed:
feat: implement Containers deduplication

TODO:
- decrease code duplication
- add tests

cmltaWt0 committed:
test: fix modulestore_migrator test

Changes:
- fix updating the existing_source_to_target_keys dictionary process
- fix new slug generation for Contaners

cmltaWt0 committed:
feat: implement MigrationContext

Changes:
- refactor migration context usage
- update tests
- add new tests
- fix typo in return type for _deduplicate_container

cmltaWt0 committed:
refactor: avoid redundant DB queries

cmltaWt0 committed:
fix: Rest API create action returns incorrect value

cmltaWt0 committed:
refactor: update Rest API contract

Changes:
- remove `name` field
- rename `modulestoremigration` to `parameters`
- update docstring to correctly describe `state_text`

kdmccormick commented on cms/djangoapps/modulestore_migrator/migrations/0001_initial.py:
```diff
@@ -0,0 +1,106 @@
+import opaque_keys.edx.django.models
+
+
+class Migration(migrations.Migration):
```
Could you squash these migrations down into one?

kdmccormick committed:
fix: remove unnecessary imports left over from rebase

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+
+
+@dataclass(frozen=True)
+class MigrationContext:
```
Nice refactoring. This could include content_by_filename as well, right?

cmltaWt0 replied:
I wasn't sure about this one. Will add then.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+
+@dataclass(frozen=True)
+class MigrationContext:
+    existing_source_to_target_keys: dict[UsageKey, PublishableEntity]
```
Could you add a note that this field gets mutated during the import process? Since the rest of the data is frozen, a dev might assume that this field is kept constant as well.

kdmccormick replied:
could you adjust the spacing or wording so that it's clear that the note only applies to `existing_source_to_target_keys`, not to the other fields?

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+        for block_source, block_migration in block_sources_to_block_migrations.items():
+            block_source.forwarded = block_migration
+            block_source.save()
+        # ModulestoreBlockSource.objects.bulk_update(block_sources_to_block_migrations.keys(), ["forwarded"])
```
```suggestion
```

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+            # time consumption during this step
+            # For existing and deleted blocks this step takes
+            # up to 110 seconds
+            # For new blocks this step takes up to 3 seconds
```
Can you write a GitHub issue for this and link it here?

cmltaWt0 replied:
Can't reproduce it yet. Want to run other use cases to try to get the issue.

cmltaWt0 replied:
The TODO is removed - can't reproduce the issue.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+    #                                     releases may support treating the Course as an importable container).
+    #   * Something else              --> Try to import it as a component. If that fails, then it's either an un-
+    #                                     supported component type, or it's an XBlock with dynamic children, which we
+    #                                     do not support in libraries as of Teak.
```
```suggestion
    Migrate an OLX node (source_node) from a legacy course or library (source_context_key) to a
    learning package (context.target_library). If the node is a container, create it in the target if
    it is at or above the requested composition_level; otherwise, just import its contents.
    Recursively apply the same logic to all children.
    """
    # The OLX tag will map to one of the following...
    #   * A wiki tag                  --> Ignore
    #   * A recognized container type --> Migrate children, and import container if requested.
    #   * A legacy library root       --> Migrate children, but NOT the root itself.
    #   * A course root               --> Migrate children, but NOT the root itself (for Ulmo, at least. Future
    #                                     releases may support treating the Course as an importable container).
    #   * Something else              --> Try to import it as a component. If that fails, then it's either an un-
    #                                     supported component type, or it's an XBlock with dynamic children, which we
    #                                     do not support in libraries as of Ulmo.
```

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+                #
+                # @@TODO: Implement a stricter enforcement mechanism.
+                # The system should prevent a node from being migrated without an update
+                # to the `existing_source_to_target_keys` dictionary.
```
Seems OK to me if we don't have a stricter enforcement mechanism. If MigrationContext were meant to be used across different edx-platform apps, then it'd be important to be more strict. But since it just is used for tasks.py, I think what you have now is fine.

kdmccormick replied:
Remember to remove this `@@TODO`

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+
+
+@dataclass(frozen=True)
+class MigrationContext:
```
```suggestion
class _MigrationContext:
```
Let's mark this as private, as it's not meant to be used outside tasks.py.

cmltaWt0 replied:
Agreed. I actually hesitated to mark it private and also tend to use private variables for this module.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+) -> LibraryContainerLocator:
+    if not source_key or not source_key.block_id:
+        raise ValueError(f"Invalid source_key: {source_key}")
+
```
```suggestion
```
Since we're using type annotations, mypy should forbid this from ever happening, so I think it we can remove this case. Unless you were seeing it happen in your testing?

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+    source_key: UsageKey,
+    container_type: ContainerType,
+    title: str,
+) -> LibraryContainerLocator:
```
Can you add a docstring for this method?

Also, I would suggest a different names. "deduplicate" means when you have multiple instances of the same thing, so you merge them all into one instance. This is kinda the opposite: we have multiple things (containers) and we want to find a key for each one so that they all remain separate. Maybe `_get_distinct_target_container_key` and `_get_distinct_target_usage_key`?

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+        """Update the context with a new migration (keeps it current)"""
+        self.existing_source_to_target_keys[source_key] = target
+
+    def get_migrated_entities(self, base_key: str) -> set[UsageKey]:
```
```suggestion
    def get_existing_target_entity_keys(self, base_key: str) -> set[str]:
```
We're returning entity keys (which are strings), so let's update the name and type signature to match that.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+                raise ValueError(f"Invalid key format: {existing_target.key}")
+            block_id = key_parts[-1]
+        except (AttributeError, IndexError) as e:
+            raise ValueError(f"Failed to parse existing key: {existing_target.key}") from e
```
I'm unsure if we should be doing direct string manipulation of these keys. As you're familiar with, when we work with OpaqueKeys (like CourseKeys) it's best practice to use the objects methods (`course_key.run`, etc.) instead of assuming the key format. But entity keys are just strings, so I'm not sure how we're supposed to safely work with them. These entity keys are new to the platform and I'm not sure if we have an established best practice yet... let me do some research and come back to this comment.

kdmccormick replied:
@cmltaWt0 OK, I think the safest thing to do here would be load the related component and look at its `local_key`, which is guaranteed to always contain the block id:

```suggestion
        block_id = existing_target.component.local_key
```

It adds up to one query for each existing block, but I think this is worth the robustness, and if we see performance issues we can address it in the future.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+    slug = slugify(title, allow_unicode=True)
+    if slug:
+        if use_hash:
+            return f"{slug}-{uuid4().hex[-6:]}"
```
Do we need use_hash as an option?

cmltaWt0 replied:
Removed.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+        else _slugify_source_usage_key(
+            source_key,
+            title,
+        )
```
Presuming we can remove the `use_hash` option (see below), I think `_slugify_source_usage_key` is now so simple that it'd be more readable just to inline it.
```suggestion
        source_key.block_id
        if context.preserve_url_slugs
        else (slugify(title) or source_key.block_id)
```

cmltaWt0 replied:
Agreed. I wasn't sure whether we want the hash or not.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+# Not used for now since it causes issues with
+# adding to Collections
+# Required for additional investigation
+def _create_migration_artifacts_in_bulk(
```
If we're going to leave this in the PR, then please create follow-up issue and link it here to help prevent us from forgetting about it and leaving dead code here.

cmltaWt0 replied:
Can't reproduce the issue. Will test additional use cases.
@kdmccormick What do you think will be better - to leave two different function implementing a third coordinator func to invoke a desired logic based on input param (e.g. create_in_bulk=True) and simple leave one option.
I'm not sure on the bulk option since it's really complicated and obviously harder to maintain than incremental one. And the performance boost for regular courses is not so huge.

kdmccormick replied:
My preference is to never add dead code, so let's just keep one implementation (the incremental one). But in order to keep the bulk-create code available in case we want to reinstate it later, can you just paste this function as a code-block in a comment on https://github.com/openedx/frontend-app-authoring/issues/1681 ?

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,923 @@
+        return True
+    except IntegrityError as e:
+        log.warning(f"IntegrityError creating content {key}: {e}")
+        return False
```
Why do we need to catch IntegrityError?

cmltaWt0 replied:
Also not sure about this. I tend to completely remove this func since the `create_component_version_content` already do `get_or_create` and we do not need to check a component version for a presence separately.

kdmccormick replied:
Agreed

kdmccormick reviewed (commented):
@cmltaWt0 Thanks for all your work. It's great to see the conflict resolution working. I think we will need a couple more review passes, but I think we are close to merging the first version of this, and then we can follow up later with performance fixes.

kdmccormick commented on cms/djangoapps/modulestore_migrator/models.py:
```diff
@@ -0,0 +1,216 @@
+        help_text=_(
+            "If a piece of content already exists in the content library, should the import process replace it?"
+        ),
+    )
```
As we discussed in the sync today, we'll actually need to support a third mode in the future: `create_copy`. We don't need to implement that now, we can implement it in the future: https://github.com/openedx/frontend-app-authoring/issues/2378. But, in order to make this PR's data model compatible, can you change this to be ChoiceField? I'm thinking the field could be called `repeat_handling_strategy`, with choices `update`, `skip`, and `create_copy`. For now, we'll just raise a NotImplementedError on `create_copy`.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11043825783.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11043825783-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11043825783-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11043825783-tutor-requirements.txt)

kdmccormick commented on cms/djangoapps/modulestore_migrator/rest_api/v0/urls.py:
```diff
@@ -0,0 +1,11 @@
+from .views import ImportViewSet
+
+ROUTER = SimpleRouter()
+ROUTER.register(r'imports', ImportViewSet)
```
Rather than `/api/modulestore_migrator/v0/imports/`, can we put the API under `/api/modulestore_migrator/v1/migrations/`?

Reasoning:
* This API needs to be stable in Ulmo in order to support the legacy libraries migration, so V1 is appropriate.
* The rest of the app calls it "migrations" instead of "imports", so let's be consistent with that in the code, even thought the UI will say "Import".

cmltaWt0 committed:
feat: implement RepeatHandlingStrategy

Implementated as part of https://github.com/openedx/frontend-app-authoring/issues/2378
Allows for a future work on Fork strategy.

kdmccormick commented:
I've written down the latest API schema so that OpenCraft can soon use it to begin implementing the legacy libraries migration frontend: https://openedx.atlassian.net/wiki/spaces/OEPM/pages/5173182480/REST+API+for+modulestore_migrator

I believe this PR will implement everything in that page except for:
* the `?sources=` filter key, which I propose we make a follow-up PR
* the ability for course authors to start and to see migrations for their courses and libraries, rather than requiring that a global admin do it.

navinkarkera commented:
@kdmccormick I am working on the API to [filter migrated legacy library blocks](https://github.com/openedx/frontend-app-authoring/issues/2172). How would we know whether a source (a legacy library in this case) has been completed migrated to v2?

kdmccormick commented:
@navinkarkera are you looking for a REST API, or are you building a REST API and so you need a Python API?

navinkarkera commented:
@kdmccormick I am building the REST API, so I need a python API. I can even build the python API for it but not sure about differentiating between partially migrated and completed migrated legacy sources.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11129151332.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11129151332-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11129151332-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11129151332-tutor-requirements.txt)

kdmccormick commented:
@navinkarkera Gotcha. There is no Python API yet, but to answer the question "is a legacy library successfully migrated?" you could write something like this:

```python
def is_successfully_migrated(source_key: CourseKey | LibraryLocator) -> bool:
    ModulestoreSource.get_or_create(
        key=str(source_key)
    )[0].migrations.filter(state=UserTaskStatus.SUCCEEDED).exists()
```

Do you need something more granular / block-level? Depending on what "partial import" means, we could query further down (to ModulestoreBlockSource and ModulestoreBlockMigration), or perhaps the migration tool needs to aggregate some more information at the ModulestoreMigration level. In case, could you bring the discussion to another issue so we can focus on PR review here?

cmltaWt0 committed:
refactor: address code review comments, clean code

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11140161925.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11140161925-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11140161925-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11140161925-tutor-requirements.txt)

cmltaWt0 commented:
@kdmccormick Could you check on my rights to force push to the source branch? We need to rebase and resolve conflicts.
I've left two comments for Bulk ModulestoreBlockMigration creation and for performance issues. All other comments seems to be resolved.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11142707209.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11142707209-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11142707209-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11142707209-tutor-requirements.txt)

kdmccormick commented:
@cmltaWt0 Yes, feel free to force push! Or is it not allowing you to?

cmltaWt0 committed:
refactor: remove _create_migration_artifacts_in_bulk

cmltaWt0 committed:
fix: imports after rebase

cmltaWt0 committed:
style: improve code style quality

kdmccormick commented on cms/djangoapps/modulestore_migrator/models.py:
```diff
@@ -0,0 +1,217 @@
+        ModulestoreSource,
+        on_delete=models.CASCADE,
+        related_name="migrations",
+    )
```
@cmltaWt0 , I just realized that it would be very useful to have a `source_version` CharField here so that we can know what version a course/library was when it was migrated. For legacy libraries in particular, it would be good to know the source version so that we can know whether a "sync" is available for migrated legacy library references. Could you add that field and populate it during the migration process?

cmltaWt0 replied:
Absolutely! Great idea.

cmltaWt0 replied:
Pushed the quick implementation (use `source.key.CANONICAL_NAMESPACE` for locator version).
Could you review this?

cmltaWt0 committed:
feat: add source_version field for ModulestoreMigration model

Changes:
- update initial migration
- update tests

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,781 @@
+    # For legacy libraries in particular, it would be good to know
+    # the source version so that we can know whether a "sync" is
+    # available for migrated legacy library references
+    source_version = source.key.CANONICAL_NAMESPACE
```
@cmltaWt0 This is the version of the key scheme (legacy `library-v1` vs new `lib`), what I want to have is rather the version of the content itself. Specifically, every modulestore course has a version hash which changes every time any block in it is edited. I forget the exact field name, but I think it will be available on the `legacy_root` block below. Or, if it is not available on that, it will be associated with the course object which you can load using `modulestore().get_course` on the source key.

If it does not seem trivial to add this now, please make it a follow-up PR for after we merge this one.

cmltaWt0 replied:
Thanks for the clarification. I'll try to come up with a proper solution.

cmltaWt0 replied:
I think it won't be hard to implement. I've just thought about different use case :)

cmltaWt0 replied:
I believe we can to use the `version_guid` for this.
Library: `library.location.library_key.version_guid`
Course: `course.course_key.version_guid`

kdmccormick replied:
Yes, sounds right! Just be aware that sometimes version_guid is None, depending on how you load the block from modulestore. I think that version_guid is always set to a real value when you load a block from CMS (it's None by-default LMS). But I think it would be good to check for None and fail if it's None here just to be sure.

cmltaWt0 replied:
Honestly, I've played with this approach for a while and never had anything except the None.
For now - I've implemented it via `SplitModulestoreCourseIndex` and it works pretty stable. Please review - for me it looks ok but maybe I miss something and have to switch back to `version_guid`.

cmltaWt0 replied:
There is what I have right now.
<img width="539" height="474" alt="Screenshot 2025-09-04 at 16 15 14" src="https://github.com/user-attachments/assets/945d1bab-b689-4dda-8f04-d83afd2c504e" />
<img width="791" height="395" alt="Screenshot 2025-09-04 at 16 14 59" src="https://github.com/user-attachments/assets/09ff7bdd-cd89-468e-9b70-b28aec3dfbe1" />
<img width="814" height="352" alt="Screenshot 2025-09-04 at 16 14 46" src="https://github.com/user-attachments/assets/1ecdbe42-b0cc-4386-b4a4-040be8988464" />

kdmccormick replied:
perfect!

kdmccormick commented on cms/djangoapps/modulestore_migrator/rest_api/v1/views.py:
```diff
@@ -0,0 +1,137 @@
+                composition_level=validated_data['composition_level'],
+                repeat_handling_strategy=validated_data['repeat_handling_strategy'],
+                preserve_url_slugs=validated_data['preserve_url_slugs'],
+                forward_source_to_target=False,  # @@TODO - Set to False for now. Explain this better.
```
@cmltaWt0 can you set this from the serializer data and remove the comment?

```suggestion
                forward_source_to_target=False,  # @@TODO - Set to False for now. Explain this better.
```

cmltaWt0 replied:
@kdmccormick Do we need this param in response data?
```json
{
    "artifacts": [],
    "parameters": {
        "source": "ModulestoreSource('library-v1:RG+lib-legacy')",
        "target": "lib:RG:test19",
        "composition_level": "component",
        "repeat_handling_strategy": "update",
        "preserve_url_slugs": false,
        "forward_source_to_target": true
    }
}
```

Asking because there is no such field and I need either to pass this param directly to response or implement custom logic to get the non existent field.

Right now it can be as simple as this line in view:
```python
serializer.data["parameters"]["forward_source_to_target"] = validated_data['forward_source_to_target']
```

kdmccormick replied:
ah good call out.

We need this to be returned from the GET as well, so that would mean adding it to the serializer with custom logic, right? If so, please do that.

kdmccormick replied:
^ Reminder on this.

I think the logic could be something like this:

```python
def get_forward_source_to_target(self, obj: ModulestoreMigration):
    return obj.id == obj.source.forwarded_id
```

kdmccormick commented:
@cmltaWt0 Ready for final review?

cmltaWt0 commented:
> @cmltaWt0 Ready for final review?

I'm removing TODOs, resolving the `forward_source_to_target` thing and fixing linter complains.
Besides this - it's ready for the final review.

cmltaWt0 committed:
feat: use forward_source_to_target from request data for REST API

Other changes:
- clean code
- remote TODOs

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11249846755.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11249846755-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11249846755-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11249846755-tutor-requirements.txt)

kdmccormick commented on cms/djangoapps/modulestore_migrator/models.py:
```diff
@@ -0,0 +1,223 @@
+        ),
+    )
+    preserve_url_slugs = models.BooleanField(
+        default=True,
```
```suggestion
        default=False,
```
Let's default to False, since existing URL slugs are mostly bad :)

kdmccormick commented on cms/djangoapps/modulestore_migrator/models.py:
```diff
@@ -0,0 +1,223 @@
+    preserve_url_slugs = models.BooleanField(
+        default=True,
+        help_text=_(
+            "If imported blocks have hashed slugs - do we want to preserve it?"
```
The code doesn't care if the existing URL slugs are hashed or not; rather it always just listens to the value of `preserve_url_slugs`, correct?

Assuming so, please change to `Should the migration preserve the location IDs of the existing blocks? If not, then new, unique human-readable IDs will be generated based on the block titles.`

cmltaWt0 replied:
It's correct. Will update the help text.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,774 @@
+        ]
+    source_to_target: tuple[UsageKey, PublishableEntityVersion] | None = None
+    if should_migrate_node:
+        source_olx = etree.tostring(source_node).decode('utf-8')
```
The `_get_title_from_olx` function has some issues:

* it's importing a private function (`_title_from_olx_node`)
* it's unnecessarily re-parsing each component's OLX
* it's special logic for components; containers are handled differently right here.

Rather than have the helper function, I think we could keep it simple and use the same inline logic to compute titles for both containers and components, passing the titles down into `_migrate_component`:
```python
        title = source_node.get('display_name', source_block_id()
        ....
        ....
                _migrate_component(
                    context=context,
                    source_key=source_key,
                    olx=source_olx,
                    title=title,
                )
```

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -0,0 +1,774 @@
+    Migrate an OLX node (source_node) from a legacy course or library (source_context_key) to a
+    learning package (context.target_library). If the node is a container, create it in the target if
+    it is at or above the requested composition_level; otherwise, just import its contents.
+    Recursively apply the same logic to all children.
```
source_context_key -> context.source_context_key
```suggestion
    Migrate an OLX node (source_node) from a legacy course or library (context.source_context_key)
    to a learning package (context.target_library). If the node is a container, create it in the
    target if it is at or above the requested composition_level; otherwise, just import its contents.
    Recursively apply the same logic to all children.
```

kdmccormick commented on cms/djangoapps/modulestore_migrator/rest_api/v1/views.py:
```diff
@@ -0,0 +1,140 @@
+log = logging.getLogger(__name__)
+
+
+class ImportViewSet(StatusViewSet):
```
```suggestion
class MigrationViewSet(StatusViewSet):
```
Although this will be used in the "Import Course to Library" user interface, let's stick with "migration" as the term we use in the API and the backend.

kdmccormick reviewed (commented):
Just a few more minor comments, but otherwise the whole changeset looks solid to me.

I tested several different migrations myself (using the admin, not the RESTAPI) and it works great! It performs surprisingly well, too-- it took under a minute to migrate the entire demo course on my local machine. Great work 🚀 

If you are busy this week @cmltaWt0 , I have some spare cycles and would not mind using them to resolve these comments and fix the linter. Let me know if that would be helpful.

cmltaWt0 commented:
> Just a few more minor comments, but otherwise the whole changeset looks solid to me.
> 
> I tested several different migrations myself (using the admin, not the RESTAPI) and it works great! It performs surprisingly well, too-- it took under a minute to migrate the entire demo course on my local machine. Great work 🚀
> 
> If you are busy this week @cmltaWt0 , I have some spare cycles and would not mind using them to resolve these comments and fix the linter. Let me know if that would be helpful.

I'm going to resolve some or all comments today.

cmltaWt0 commented:
Regarding this comment: https://github.com/openedx/edx-platform/pull/36873#discussion_r2333846490
The logic to determine it is clear. The issue is in the field - `get_forward_source_to_target` method is not invoked for writable fields. So I need to either change the field or add custom serialization.
Actually I've added a custom serialization via `to_representation` method and removed the logic from view.

cmltaWt0 committed:
fix: resolve import linting complains

Get content library directly from learning package.

cmltaWt0 commented:
@kdmccormick I've addressed the import linting error in this [commit](https://github.com/openedx/edx-platform/pull/36873/commits/cdd380c5c3bbba1b817276631469f12f5e731852). Since you wrote the original code, could you please take a look?

I'm not entirely happy with my approach, as I had to access the model via its reverse relation. I couldn't find a more direct way through the content_libraries/api.py module, other than constructing a LibraryLocatorV2 and calling the get_library function. If you know of a better way, I'd appreciate your feedback!

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11309643573.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11309643573-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11309643573-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11309643573-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11310061182.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11310061182-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11310061182-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11310061182-tutor-requirements.txt)

kdmccormick commented:
> other than constructing a LibraryLocatorV2 and calling the get_library function

Hmm, I think this is the best way.

cmltaWt0 commented:
> > other than constructing a LibraryLocatorV2 and calling the get_library function
> 
> Hmm, I think this is the best way.

So you prefer to use `get_library` instead of reverse relation access?

kdmccormick commented:
Yeah

cmltaWt0 committed:
refactor: use content_libraries.api.get_library to get a library in a task

Added new `target_library_key` arg to pass into migrate_from_modulestore
task to re-construct the target library.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401258.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401258-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401258-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401258-tutor-requirements.txt)

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401263.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401263-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401263-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11329401263-tutor-requirements.txt)

kdmccormick commented:
@cmltaWt0 Ready for me to review and merge? Don't worry about the sandbox build, I've tested locally, and I presume you have too.

cmltaWt0 commented:
> @cmltaWt0 Ready for me to review and merge? Don't worry about the sandbox build, I've tested locally, and I presume you have too.

Definitely ready for review. But I would like to update the branch and test it in `local` mode a bit.

kdmccormick commented:
Sounds good @cmltaWt0 . If you would like the sandbox back for testing, then put the `create-sandbox` label back and it should come up within an hour. Just be sure to rebase--it was failing before because the branch was out of date.

open-craft-grove commented:
Sandbox deployment failed 💥
Please check the settings and requirements.
Retry deployment by pushing a new commit or updating the requirements/settings in the pull request's description.
📜 [Failure Logs](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11339479426.log)
ℹ️ [Grove Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11339479426-tutor-config.yml), [Tutor Config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11339479426-grove-config.yml), [Tutor Requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-36873-7daceb-51564656-11339479426-tutor-requirements.txt)

kdmccormick commented:
@cmltaWt0 asked:

> Testing the REST API with real Celery (in local mode) and I feel like I need to recall the desired behaviour in this case.
> 1. When the REST API is about to response to client we do not have a running task yet.
> 2. So we do not have a Migration object to serialize and the response contains “parameters” = null
> 3. And API do not contain the task UUID to fetch later.
> 
> I’m wondering whether I miss it somewhere or it’s ok for now as a basis for a future work.

kdmccormick commented:
Good catch, that is indeed not the intended behavior.

Would it be feasible to do this @cmltaWt0 ?

1. When starting the celery task, a Task object is returned (I believe), which contains metadata about the task.
2. Query the UserTaskStatus table to get the object associated with that Task-- this will provide you the UUID to return in the response.
3. For the "parameters", I think you can construct an unsaved `Migration(...)` model object and serialize it.

(1) and (2) are important for this PR, otherwise it will be impossible to check back on the status of the task.

(3) is just a nice-to-have; it's not critical. If it is not trivial to do, then I would prefer that you omit `parameters` from the POST response. In that case, I can either make a follow-up ticket, or just remove it from the API.

cmltaWt0 commented:
> Good catch, that is indeed not the intended behavior.
> 
> Would it be feasible to do this @cmltaWt0 ?
> 
>     1. When starting the celery task, a Task object is returned (I believe), which contains metadata about the task.
> 
>     2. Query the UserTaskStatus table to get the object associated with that Task-- this will provide you the UUID to return in the response.
> 
>     3. For the "parameters", I think you can construct an unsaved `Migration(...)` model object and serialize it.
> 
> 
> (1) and (2) are important for this PR, otherwise it will be impossible to check back on the status of the task.
> 
> (3) is just a nice-to-have; it's not critical. If it is not trivial to do, then I would prefer that you omit `parameters` from the POST response. In that case, I can either make a follow-up ticket, or just remove it from the API.

Working on (1) and (2). In case (3) will be really quick and easy - add it to a new commit too.


UPD:
1. Adding `uuid` is pretty simple - I missed that the default `StatusSerializer` [excludes](https://github.com/openedx/django-user-tasks/blob/master/user_tasks/serializers.py) the field from the list and we just have to manually add it (https://github.com/openedx/edx-platform/pull/36873/commits/c8725fb3eea68ec80c5391e151992cd2685f7ae8#diff-2946df10d998cf2b4006089283ce2cc088149099d1e2d60f44fc8d1369574fe7R118).
2. Adding `parameters`: won't do for now due to lack of time.

@kdmccormick fyi

cmltaWt0 committed:
fix: add missed uuid field to migrator REST API response

Default StatusSerializer explicitly excludes the uuid field so
we have to add it to StatusWithModulestoreMigrationSerializer

rpenido commented on cms/djangoapps/modulestore_migrator/api.py:
```diff
@@ -0,0 +1,57 @@
+    repeat_handling_strategy: str,
+    preserve_url_slugs: bool,
+    forward_source_to_target: bool,
+) -> None:
```
Nit: we are returning `AsyncResult` from `celery.result` here
```suggestion
) -> AsyncResult:
```

cmltaWt0 replied:
@rpenido Updated. Thanks!

rpenido commented:
Hi @kdmccormick! Is there something we could do to help get this merged?
We have some other PRs depending on this:
- https://github.com/openedx/edx-platform/pull/37286
- https://github.com/openedx/frontend-app-authoring/pull/2417
- https://github.com/openedx/frontend-app-authoring/pull/2421
- https://github.com/openedx/frontend-app-authoring/pull/2413

Thanks! :smiley:

cmltaWt0 committed:
style: update return type hint for start_migration_to_library

cmltaWt0 commented:
> Hi @kdmccormick! Is there something we could do to help get this merged? We have some other PRs depending on this:
> 
>     * [feat: updates legacy libraries list API to include migration info [FC-0097] #37286](https://github.com/openedx/edx-platform/pull/37286)
> 
>     * [feat: show migration status in libraries list [FC-0097] frontend-app-authoring#2417](https://github.com/openedx/frontend-app-authoring/pull/2417)
> 
>     * [feat: migration filter and search bar in legacy libraries tab [FC-0097] frontend-app-authoring#2421](https://github.com/openedx/frontend-app-authoring/pull/2421)
> 
>     * [feat: add library v2 alert [FC-0097] frontend-app-authoring#2413](https://github.com/openedx/frontend-app-authoring/pull/2413)
> 
> 
> Thanks! 😃

@rpenido Do you have any additional suggestions and concerns regarding the REST API contract or behavior that I can update before merge?

The only subtle think I can imagine right now is `forward_source_to_target` field. It's calculated dynamically `return obj.id == obj.source.forwarded_id` and it's became true when we actually have `source.forwarded_id.
So it's became true at the latest STEPS of the migration process - so basically if we get the info about the migration we get `false` for several calls.

@kdmccormick despite this I believe we can merge it.

rpenido commented:
> Do you have any additional suggestions and concerns regarding the REST API contract or behavior that I can update before merge?

Nothing for now. You did a good job here! :smiley:

kdmccormick commented:
> The only subtle think I can imagine right now is forward_source_to_target field. It's calculated dynamically return obj.id == obj.source.forwarded_id and it's became true when we actually have source.forwarded_id. So it's became true at the latest STEPS of the migration process - so basically if we get the info about the migration we get false` for several calls.

Ah, you're right, that is a bug, and we should fix it before Ulmo. But I agree that we can still merge this PR now.

Can you open a followup ticket for this?

kdmccormick reviewed (commented):
Approved! ✅ (GH won't let me approve my own PR)

Thank you for all the hard work @cmltaWt0 .

cmltaWt0 commented:
> > The only subtle think I can imagine right now is forward_source_to_target field. It's calculated dynamically return obj.id == obj.source.forwarded_id and it's became true when we actually have source.forwarded_id. So it's became true at the latest STEPS of the migration process - so basically if we get the info about the migration we get false` for several calls.
> 
> Ah, you're right, that is a bug, and we should fix it before Ulmo. But I agree that we can still merge this PR now.
> 
> Can you open a followup ticket for this?

Sure thing, will do.
